custom_components/octopus_intelligent/octopus_intelligent_system.py

```diff
@@ -33,30 +34,110 @@ def __init__(self, hass, *, api_key, account_id, off_peak_start, off_peak_end):
 
-    async def _async_update_data(self):
+    @override
+    async def _async_update_data(self) -> dict[str, Any]:
         """Fetch data from API endpoint.
 
         This is the place to pre-process the data to lookup tables
         so entities can quickly look up their data.
+
+        Returns:
+            dict: The data received from the Octopus API, for example:
+            {
+                'completedDispatches': [{
+                    'chargeKwh': '-0.58',
+                    'startDtUtc': '2024-02-25 02:00:00+00:00',
+                    'endDtUtc': '2024-02-25 02:30:00+00:00',
+                    'meta': {'location': 'AT_HOME', 'source': None},
+                }, {
+                    'chargeKwh': '-0.58',
+                    'startDtUtc': '2024-02-25 03:30:00+00:00',
+                    'endDtUtc': '2024-02-25 04:00:00+00:00',
+                    'meta': {'location': 'AT_HOME', 'source': None},
+                }],
+                'plannedDispatches': [{
+                    'chargeKwh': '-0.67',
+                    'startDtUtc': '2024-02-25 23:30:00+00:00',
+                    'endDtUtc': '2024-02-26 00:00:00+00:00',
+                    'meta': {'location': None, 'source': 'smart-charge'},
+                }, {
+                    'chargeKwh': '-1.12',
+                    'startDtUtc': '2024-02-26 03:00:00+00:00',
+                    'endDtUtc': '2024-02-26 04:00:00+00:00',
+                    'meta': {'location': None, 'source': 'smart-charge'},
+                }],
+                'registeredKrakenflexDevice': {
+                    'chargePointMake': 'No charger (3-pin plug)',
+                    'chargePointModel': '3-pin plug',
+                    'chargePointPowerInKw': '2.400',
+                    'createdAt': '2024-02-08T10:43:34.061834+00:00',
+                    'hasToken': True,
+                    'krakenflexDeviceId': '...',
+                    'provider': 'ENODE',
+                    'status': 'Live',
+                    'suspended': False,
+                    'vehicleBatterySizeInKwh': '58.00',
+                    'vehicleMake': 'Volkswagen',
+                    'vehicleModel': 'ID.3 Pro',
+                },
+                'vehicleChargingPreferences': {
+                    'weekdayTargetSoc': 80,
+                    'weekdayTargetTime': '08:00',
+                    'weekendTargetSoc': 80,
+                    'weekendTargetTime': '08:00',
+                },
+            }
         """
         try:
             # Note: asyncio.TimeoutError and aiohttp.ClientError are already
             # handled by the data update coordinator.
-            async with async_timeout.timeout(90):
-                return await self.client.async_get_combined_state(self._account_id)
+            async with asyncio.timeout(90):
+                data = await self.client.async_get_combined_state(self._account_id)
+                self._update_planned_dispatch_sources(data)
+                return data
```

Comment:
As @matt7aylor had [pointed out](https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197) in issue #35, sometimes the planned dispatch list gets changed from `"source": "smart-charge"` to `"source": None`. The idea here is to remember the last non-None source, save it in `self._last_seen_planned_dispatch_source`, and then replace None with the saved source when None is found. I’ve tested this for over a week and it has been working fine for me. Example:

Initially, `'source': 'smart-charge'`:
```
2024-02-28 05:55:02.838 ...
 'plannedDispatches': [{'chargeKwh': '-7.31',
                        'endDtUtc': '2024-02-28 08:30:00+00:00',
                        'meta': {'location': None, 'source': 'smart-charge'},
                        'startDtUtc': '2024-02-28 05:20:00+00:00'}],
```
Then it goes rogue, `'source': None`:
```
2024-02-28 06:00:04.618 ...
 'plannedDispatches': [{'chargeKwh': '-5.77',
                        'endDtUtc': '2024-02-28 08:30:00+00:00',
                        'meta': {'location': None, 'source': None},
                        'startDtUtc': '2024-02-28 06:00:00+00:00'}],
```
But the code in this PR replaces `None` with `smart-charge` (in this example) before passing it on to the integration sensors etc.